compiler/rustc_lint_defs/src/builtin.rs

```diff
@@ -4272,31 +4275,104 @@ declare_lint! {
 }
 
 declare_lint! {
-    /// The `unknown_or_malformed_diagnostic_attributes` lint detects unrecognized or otherwise malformed
-    /// diagnostic attributes.
+    /// The `malformed_diagnostic_attributes` lint detects malformed diagnostic attributes.
     ///
     /// ### Example
     ///
     /// ```rust
-    /// #![feature(diagnostic_namespace)]
-    /// #[diagnostic::does_not_exist]
-    /// struct Foo;
+    /// #[diagnostic::do_not_recommend(message = "message")]
+    /// trait Trait {}
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// It is usually a mistake to use options or syntax that is not supported. Check the spelling,
+    /// and check the diagnostic attribute listing for the correct name and syntax. Also
+    /// consider if you are using an old version of the compiler; perhaps the option or syntax
+    /// is only available in a newer version. See the [reference] for a list of diagnostic attributes
+    /// and their syntax.
+    ///
+    /// [reference]: https://doc.rust-lang.org/nightly/reference/attributes/diagnostics.html#the-diagnostic-tool-attribute-namespace
+    pub MALFORMED_DIAGNOSTIC_ATTRIBUTES,
+    Warn,
+    "detects malformed diagnostic attributes",
+}
+
+declare_lint! {
+    /// The `misplaced_diagnostic_attributes` lint detects wrongly placed diagnostic attributes.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// #[diagnostic::do_not_recommend]
+    /// struct NotUserFacing;
     /// ```
     ///
     /// {{produces}}
     ///
+    /// ### Explanation
+    ///
+    /// It is usually a mistake to specify a diagnostic attribute on an item it is not meant for. For example,
+    /// `#[diagnostic::do_not_recommend]` can only be placed on trait implementations, and does nothing if placed
+    /// elsewhere. See the [reference] for a list of diagnostic attributes and their correct positions.
```

Comment:
```suggestion
    /// It is usually a mistake to specify a diagnostic attribute on an item for
    /// which it is not meant. For example, `#[diagnostic::do_not_recommend]`
    /// can only be placed on trait implementations, and does nothing if placed
    /// elsewhere. See the [reference] for a list of diagnostic attributes and
    /// the correct position of each.
```

Wrapping and some wording improvements.